File: train_model.mjs
/**
 * train_model.mjs
 * ================
 * Node.js script that reads advertising.csv, performs degree-2 polynomial
 * regression using a pure-JS least-squares solver (no Python needed),
 * and exports model_coefficients.json + model_stats.json for the Next.js app.
 *
 * Polynomial features (degree=2, no bias term):
 *  [0] TV
 *  [1] Radio
 *  [2] Newspaper
 *  [3] TV^2
 *  [4] TV*Radio
 *  [5] TV*Newspaper
 *  [6] Radio^2
 *  [7] Radio*Newspaper
 *  [8] Newspaper^2
 *
 * Uses normal equations: β = (XᵀX)⁻¹ Xᵀy  with bias column prepended.
 */

import { readFileSync, mkdirSync, writeFileSync } from "fs";

// ── Read CSV ──────────────────────────────────────────────────────────────────
const csv = readFileSync("advertising.csv", "utf-8").trim().split(/\r?\n/);
const headers = csv[0].split(",");
console.log("Columns:", headers);

const rows = csv.slice(1).map((line) => {
  const [TV, Radio, Newspaper, Sales] = line.split(",").map(Number);
  return { TV, Radio, Newspaper, Sales };
});
console.log(`Rows loaded: ${rows.length}`);

// ── Polynomial Feature Expansion ──────────────────────────────────────────────
const FEATURE_NAMES = [
  "TV", "Radio", "Newspaper",
  "TV^2", "TV Radio", "TV Newspaper",
  "Radio^2", "Radio Newspaper", "Newspaper^2",
];

function polyFeatures(tv, radio, news) {
  return [
    tv, radio, news,
    tv * tv, tv * radio, tv * news,
    radio * radio, radio * news,
    news * news,
  ];
}

// Design matrix X (with bias col=1 prepended), target y
const X = rows.map(({ TV, Radio, Newspaper }) => [1, ...polyFeatures(TV, Radio, Newspaper)]);
const y = rows.map((r) => r.Sales);
const n = rows.length;
const p = X[0].length; // 10 (bias + 9 features)

// ── Matrix Helpers ────────────────────────────────────────────────────────────
function matMul(A, B) {
  const ra = A.length, ca = A[0].length, cb = B[0].length;
  const C = Array.from({ length: ra }, () => new Array(cb).fill(0));
  for (let i = 0; i < ra; i++)
    for (let k = 0; k < ca; k++)
      for (let j = 0; j < cb; j++)
        C[i][j] += A[i][k] * B[k][j];
  return C;
}

function matTranspose(A) {
  return A[0].map((_, j) => A.map((row) => row[j]));
}

function matVecMul(A, v) {
  return A.map((row) => row.reduce((s, a, j) => s + a * v[j], 0));
}

// Gaussian elimination with partial pivoting to solve Ax = b
function solveLinear(A, b) {
  const n = b.length;
  const M = A.map((row, i) => [...row, b[i]]);
  for (let col = 0; col < n; col++) {
    let maxRow = col;
    for (let row = col + 1; row < n; row++)
      if (Math.abs(M[row][col]) > Math.abs(M[maxRow][col])) maxRow = row;
    [M[col], M[maxRow]] = [M[maxRow], M[col]];
    const pivot = M[col][col];
    if (Math.abs(pivot) < 1e-12) continue;
    for (let row = 0; row < n; row++) {
      if (row === col) continue;
      const factor = M[row][col] / pivot;
      for (let k = col; k <= n; k++) M[row][k] -= factor * M[col][k];
    }
  }
  return M.map((row, i) => row[n] / row[i]);
}

// ── Normal Equations ──────────────────────────────────────────────────────────
console.log("\nSolving normal equations β = (XᵀX)⁻¹ Xᵀy ...");
const Xt = matTranspose(X);
const XtX = matMul(Xt, X);
const Xty = matVecMul(Xt, y);
const beta = solveLinear(XtX, Xty); // [intercept, c1, c2, ..., c9]

const intercept = beta[0];
const coefficients = beta.slice(1);

console.log("\nCoefficients:");
FEATURE_NAMES.forEach((name, i) => {
  console.log(`  ${name.padEnd(20)} ${coefficients[i].toFixed(6)}`);
});
console.log(`  ${"intercept".padEnd(20)} ${intercept.toFixed(6)}`);

// ── Predictions & Metrics ─────────────────────────────────────────────────────
function predict(tv, radio, news) {
  const feats = polyFeatures(tv, radio, news);
  return intercept + feats.reduce((s, f, i) => s + f * coefficients[i], 0);
}

const yHat = rows.map(({ TV, Radio, Newspaper }) => predict(TV, Radio, Newspaper));

function mean(arr) { return arr.reduce((s, v) => s + v, 0) / arr.length; }
function r2(actual, predicted) {
  const yBar = mean(actual);
  const ssTot = actual.reduce((s, v) => s + (v - yBar) ** 2, 0);
  const ssRes = actual.reduce((s, v, i) => s + (v - predicted[i]) ** 2, 0);
  return 1 - ssRes / ssTot;
}
function rmse(actual, predicted) {
  return Math.sqrt(actual.reduce((s, v, i) => s + (v - predicted[i]) ** 2, 0) / actual.length);
}

// Train/test split (80/20, last 40 rows as test)
const splitIdx = Math.floor(n * 0.8);
const yTrain = y.slice(0, splitIdx);
const yHatTrain = yHat.slice(0, splitIdx);
const yTest = y.slice(splitIdx);
const yHatTest = yHat.slice(splitIdx);

const r2Train = r2(yTrain, yHatTrain);
const r2Test = r2(yTest, yHatTest);
const rmseTrain = rmse(yTrain, yHatTrain);
const rmseTest = rmse(yTest, yHatTest);

console.log(`\nTrain R²:  ${r2Train.toFixed(4)}`);
console.log(`Test  R²:  ${r2Test.toFixed(4)}`);
console.log(`Train RMSE: ${rmseTrain.toFixed(4)}`);
console.log(`Test  RMSE: ${rmseTest.toFixed(4)}`);

// ── Feature Importance ────────────────────────────────────────────────────────
const absCoef = coefficients.map(Math.abs);
const coefSum = absCoef.reduce((s, v) => s + v, 0);
const importance = Object.fromEntries(
  FEATURE_NAMES.map((name, i) => [name, +((absCoef[i] / coefSum) * 100).toFixed(4)])
);

// ── Data Ranges ───────────────────────────────────────────────────────────────
function colStats(col) {
  const vals = rows.map((r) => r[col]);
  return {
    min: Math.min(...vals),
    max: Math.max(...vals),
    mean: +(mean(vals).toFixed(2)),
  };
}
const dataRanges = {
  TV: colStats("TV"),
  Radio: colStats("Radio"),
  Newspaper: colStats("Newspaper"),
  Sales: colStats("Sales"),
};

// ── Diminishing Returns Curves ────────────────────────────────────────────────
function linspace(start, end, n) {
  return Array.from({ length: n }, (_, i) => start + (i / (n - 1)) * (end - start));
}

const tvMean = dataRanges.TV.mean;
const radioMean = dataRanges.Radio.mean;
const newsMean = dataRanges.Newspaper.mean;

const tvRange = linspace(0, 300, 50);
const radioRange = linspace(0, 50, 50);
const newsRange = linspace(0, 100, 50);

const diminishingReturns = {
  TV: {
    budgets: tvRange.map((v) => +v.toFixed(2)),
    sales: tvRange.map((v) => +predict(v, radioMean, newsMean).toFixed(4)),
  },
  Radio: {
    budgets: radioRange.map((v) => +v.toFixed(2)),
    sales: radioRange.map((v) => +predict(tvMean, v, newsMean).toFixed(4)),
  },
  Newspaper: {
    budgets: newsRange.map((v) => +v.toFixed(2)),
    sales: newsRange.map((v) => +predict(tvMean, radioMean, v).toFixed(4)),
  },
};

// ── Sample Predictions ────────────────────────────────────────────────────────
const samplePredictions = rows.slice(0, 5).map((row) => ({
  TV: row.TV,
  Radio: row.Radio,
  Newspaper: row.Newspaper,
  actual_sales: row.Sales,
  predicted_sales: +predict(row.TV, row.Radio, row.Newspaper).toFixed(2),
}));

// ── All training data for scatter plot ────────────────────────────────────────
const allPredictions = rows.map((row) => ({
  TV: row.TV,
  Radio: row.Radio,
  Newspaper: row.Newspaper,
  actual: row.Sales,
  predicted: +predict(row.TV, row.Radio, row.Newspaper).toFixed(2),
}));

// ── Export JSON ───────────────────────────────────────────────────────────────
mkdirSync("model_output", { recursive: true });

writeFileSync(
  "model_output/model_coefficients.json",
  JSON.stringify(
    {
      intercept,
      coefficients,
      feature_names: FEATURE_NAMES,
      degree: 2,
      include_bias: false,
      feature_map: {
        TV: 0, Radio: 1, Newspaper: 2,
        "TV^2": 3, "TV Radio": 4, "TV Newspaper": 5,
        "Radio^2": 6, "Radio Newspaper": 7, "Newspaper^2": 8,
      },
    },
    null,
    2
  )
);

writeFileSync(
  "model_output/model_stats.json",
  JSON.stringify(
    {
      model: {
        type: "Polynomial Regression",
        degree: 2,
        n_features: FEATURE_NAMES.length,
        n_training_samples: splitIdx,
        n_test_samples: n - splitIdx,
      },
      performance: {
        r2_train: +r2Train.toFixed(4),
        r2_test: +r2Test.toFixed(4),
        rmse_train: +rmseTrain.toFixed(4),
        rmse_test: +rmseTest.toFixed(4),
      },
      data_ranges: dataRanges,
      feature_importance: importance,
      diminishing_returns: diminishingReturns,
      sample_predictions: samplePredictions,
      all_predictions: allPredictions,
    },
    null,
    2
  )
);

console.log("\n✅ Model training complete!");
console.log("   model_output/model_coefficients.json");
console.log("   model_output/model_stats.json");

File: train_model.py
"""
Polynomial Regression Model Trainer
====================================
Trains a degree-2 polynomial regression on the Advertising dataset
(TV, Radio, Newspaper -> Sales) and exports model coefficients + stats
as JSON for use in the Next.js application at runtime.

Features generated (degree=2, include_bias=False):
  Index | Feature
  0     | TV
  1     | Radio
  2     | Newspaper
  3     | TV^2
  4     | TV * Radio
  5     | TV * Newspaper
  6     | Radio^2
  7     | Radio * Newspaper
  8     | Newspaper^2
"""

import json
import os
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import cross_val_score, train_test_split
from sklearn.preprocessing import PolynomialFeatures

# ── Configuration ─────────────────────────────────────────────────────────────
DATASET_PATH = "advertising.csv"
OUTPUT_DIR = "model_output"
POLY_DEGREE = 2
RANDOM_STATE = 42
TEST_SIZE = 0.2

# ── Load Data ─────────────────────────────────────────────────────────────────
print("Loading dataset...")
df = pd.read_csv(DATASET_PATH)
print(f"  Shape: {df.shape}")
print(f"  Columns: {list(df.columns)}")
print(df.describe().round(2).to_string())

X = df[["TV", "Radio", "Newspaper"]].values
y = df["Sales"].values

# ── Polynomial Feature Expansion ──────────────────────────────────────────────
print(f"\nApplying degree-{POLY_DEGREE} polynomial feature expansion...")
poly = PolynomialFeatures(degree=POLY_DEGREE, include_bias=False)
X_poly = poly.fit_transform(X)
feature_names = poly.get_feature_names_out(["TV", "Radio", "Newspaper"])
print(f"  Features ({len(feature_names)}): {list(feature_names)}")

# ── Train / Test Split ────────────────────────────────────────────────────────
X_train, X_test, y_train, y_test = train_test_split(
    X_poly, y, test_size=TEST_SIZE, random_state=RANDOM_STATE
)
print(f"\nTrain size: {len(X_train)}, Test size: {len(X_test)}")

# ── Train Model ───────────────────────────────────────────────────────────────
print("\nTraining polynomial regression model...")
model = LinearRegression()
model.fit(X_train, y_train)

# ── Evaluate ──────────────────────────────────────────────────────────────────
y_pred_test = model.predict(X_test)
y_pred_train = model.predict(X_train)

r2_test = r2_score(y_test, y_pred_test)
r2_train = r2_score(y_train, y_pred_train)
rmse_test = np.sqrt(mean_squared_error(y_test, y_pred_test))
rmse_train = np.sqrt(mean_squared_error(y_train, y_pred_train))

# Cross-validation (5-fold)
cv_scores = cross_val_score(
    LinearRegression(), X_poly, y, cv=5, scoring="r2"
)

print(f"\n  Train R²:      {r2_train:.4f}")
print(f"  Test  R²:      {r2_test:.4f}")
print(f"  Train RMSE:    {rmse_train:.4f}")
print(f"  Test  RMSE:    {rmse_test:.4f}")
print(f"  CV R² (5-fold): {cv_scores.mean():.4f} ± {cv_scores.std():.4f}")

# ── Feature Importance (absolute coefficient magnitude, normalized) ────────────
coef_abs = np.abs(model.coef_)
importance = (coef_abs / coef_abs.sum() * 100).tolist()

# ── Compute Diminishing Returns Data ─────────────────────────────────────────
# For each channel, vary that channel's budget while fixing others at their mean
tv_mean = df["TV"].mean()
radio_mean = df["Radio"].mean()
news_mean = df["Newspaper"].mean()

def predict_single(tv, radio, newspaper):
    """Run a single prediction using the trained model."""
    x = np.array([[tv, radio, newspaper]])
    x_p = poly.transform(x)
    return float(model.predict(x_p)[0])

# Generate diminishing returns curves (50 points each)
tv_range = np.linspace(0, 300, 50).tolist()
radio_range = np.linspace(0, 50, 50).tolist()
news_range = np.linspace(0, 100, 50).tolist()

tv_sales = [predict_single(tv, radio_mean, news_mean) for tv in tv_range]
radio_sales = [predict_single(tv_mean, r, news_mean) for r in radio_range]
news_sales = [predict_single(tv_mean, radio_mean, n) for n in news_range]

diminishing_returns = {
    "TV": {"budgets": [round(v, 2) for v in tv_range], "sales": [round(v, 4) for v in tv_sales]},
    "Radio": {"budgets": [round(v, 2) for v in radio_range], "sales": [round(v, 4) for v in radio_sales]},
    "Newspaper": {"budgets": [round(v, 2) for v in news_range], "sales": [round(v, 4) for v in news_sales]},
}

# ── Sample Predictions (for UI verification) ─────────────────────────────────
sample_predictions = []
for _, row in df.head(5).iterrows():
    pred = predict_single(row["TV"], row["Radio"], row["Newspaper"])
    sample_predictions.append({
        "TV": round(float(row["TV"]), 1),
        "Radio": round(float(row["Radio"]), 1),
        "Newspaper": round(float(row["Newspaper"]), 1),
        "actual_sales": round(float(row["Sales"]), 1),
        "predicted_sales": round(pred, 2),
    })

# ── Export Coefficients ───────────────────────────────────────────────────────
os.makedirs(OUTPUT_DIR, exist_ok=True)

coefficients_data = {
    "intercept": float(model.intercept_),
    "coefficients": [float(c) for c in model.coef_],
    "feature_names": list(feature_names),
    "degree": POLY_DEGREE,
    "include_bias": False,
    # Mapping for JS feature reconstruction (must match PolynomialFeatures order)
    "feature_map": {
        "TV":               0,
        "Radio":            1,
        "Newspaper":        2,
        "TV^2":             3,
        "TV Radio":         4,
        "TV Newspaper":     5,
        "Radio^2":          6,
        "Radio Newspaper":  7,
        "Newspaper^2":      8,
    }
}

with open(f"{OUTPUT_DIR}/model_coefficients.json", "w") as f:
    json.dump(coefficients_data, f, indent=2)
print(f"\nCoefficients exported → {OUTPUT_DIR}/model_coefficients.json")

# ── Export Stats ──────────────────────────────────────────────────────────────
stats_data = {
    "model": {
        "type": "Polynomial Regression",
        "degree": POLY_DEGREE,
        "n_features": len(feature_names),
        "n_training_samples": int(len(X_train)),
        "n_test_samples": int(len(X_test)),
    },
    "performance": {
        "r2_train": round(float(r2_train), 4),
        "r2_test": round(float(r2_test), 4),
        "rmse_train": round(float(rmse_train), 4),
        "rmse_test": round(float(rmse_test), 4),
        "cv_r2_mean": round(float(cv_scores.mean()), 4),
        "cv_r2_std": round(float(cv_scores.std()), 4),
    },
    "data_ranges": {
        "TV":        {"min": float(df["TV"].min()),        "max": float(df["TV"].max()),        "mean": round(float(df["TV"].mean()), 2)},
        "Radio":     {"min": float(df["Radio"].min()),     "max": float(df["Radio"].max()),     "mean": round(float(df["Radio"].mean()), 2)},
        "Newspaper": {"min": float(df["Newspaper"].min()), "max": float(df["Newspaper"].max()), "mean": round(float(df["Newspaper"].mean()), 2)},
        "Sales":     {"min": float(df["Sales"].min()),     "max": float(df["Sales"].max()),     "mean": round(float(df["Sales"].mean()), 2)},
    },
    "feature_importance": {
        name: round(imp, 4)
        for name, imp in zip(feature_names, importance)
    },
    "diminishing_returns": diminishing_returns,
    "sample_predictions": sample_predictions,
}

with open(f"{OUTPUT_DIR}/model_stats.json", "w") as f:
    json.dump(stats_data, f, indent=2)
print(f"Stats exported        → {OUTPUT_DIR}/model_stats.json")

print("\n✅ Model training complete!")
print(f"   Test R²:   {r2_test:.4f}")
print(f"   Test RMSE: {rmse_test:.4f} (units: thousand dollars)")
